### DISCUSSION TITLE
Labs Feedback: @lit-labs/scoped-registry-mixin

### DISCUSSION THREAD

**Discussion by kevinpschaaf** [lit/lit#3364] [Labs Feedback]
The `@lit-labs/scoped-registry-mixin` package provides a mixin for `LitElement` that allows a component to choose the custom element tag name for child components it renders in its shadow root, to avoid tag name collisions in the global registry. The implementation is aligned to the [Scoped CustomElementRegistry](https://github.com/WICG/webcomponents/issues/716) spec proposal; however since that spec is not yet shipped in browsers, it must be used with a scoped CustomElementRegistry [polyfill](https://github.com/webcomponents/polyfills/tree/master/packages/scoped-custom-element-registry).

We want to collect your feedback on the direction of the `@lit-labs/scoped-registry-mixin` package as we iterate on it according to our Lit Labs process and move the package towards possible graduation or retirement.

- Status: **Prototyping**
- Major known issues:
  - https://github.com/lit/lit/issues/1766
  - https://github.com/lit/lit/issues/2363
- Issues: [[labs/scoped-registry-mixin]](https://github.com/lit/lit/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+%5Blabs%2Fscoped-registry-mixin%5D+)
- Docs: [README](https://github.com/lit/lit/tree/main/packages/labs/scoped-registry-mixin#lit-labsscoped-registry-mixin)
- npm: [@lit-labs/scoped-registry-mixin](https://www.npmjs.com/package/@lit-labs/scoped-registry-mixin)

Please try `@lit-labs/scoped-registry-mixin` and tell us about your experience. Leave comments with your thoughts; highlight what issues or missing features you're blocked on; give a link to a project that uses it; etc.

And so we can have some numbers to look at, you can add reactions to this comment to indicate your experience with this package:

🚀 I have tried this package and it works great. I need it. Ship it!
👍 I have tried this package and it seems like a good idea.
😕 I have tried this package and would use it, but it has serious bugs or missing features.
👀 I am interested, but haven't tried the package yet.
👎 I'm not interested in this package.

Note: other reactions like 😄, 🎉, and ❤️ don't mean anything as part of this scheme. You probably mean, and please use, 🚀, 👍, or  👀.

**Comment by vekunz** [comment]
Hey, I'm glad that Lit is looking into the Scoped Custom Element Registry topic. However, I'm not sure if I fully understand this package yet. What we would need is a way for a Lit-Element not to register in the global registry (`window. customElements `), but instead in a scoped Custom Element Registry. This registry could belong to another Lit element or, for example, to a Svelte or Vue component that is rendered as a custom element.

Is the labs package `@lit-labs/scoped-registry-mixin` the right place for this?

We use Lit in our company to build a component library that is used across many teams (number constantly growing). When a Lit element (e.g., button, checkbox, textfield, etc.) is used by multiple teams, we run into the issue that each custom element name can only be registered once. To work around this, we currently have a small "hack" on every page where we override the global `window.customElements.define` function to prevent runtime exceptions. But even with this, we could run into version conflicts.

**Comment by sorvell** [comment]
@vekunz 

> Is the labs package @lit-labs/scoped-registry-mixin the right place for this?

This package is about scoping the elements used within a hosting LitElement itself. For example, let's say a HostElement needs to render and use a `child-element` (MyChildElement) inside itself. This package facilitates HostElement having its own scoped registry in which it can use `child-element` without any collision concerns.  How HostElement itself is defined is up to its developer and user and not part of this package.

A good general pattern is to provide a side-effect free module that exports an element class separate from one that defines the element. That way the class can be imported and easily used in a scoped registry, either in conjunction with this package or manually.

> This registry could belong to another Lit element or, for example, to a Svelte or Vue component that is rendered as a custom element.

I just want to note that frameworks like Svelte and Vue will likely need to add specific support for scoped registries. This is because creating elements with specific registries is opt-in: `document.createElement(tagName, {customElementRegistry})`...

**Reply by vekunz** [reply]
For Svelte I started a disussion there too https://github.com/sveltejs/svelte/discussions/18148.

**Comment by LorenzKahl** [comment]
Scoped Custom Element Registries (SCER 😉) is in [interop 2026](https://wpt.fyi/interop-2026) and Firefox 150 has just shipped basic support for it **behind a flag**. So hopefully SCERs will be baseline really soon (like next year-ish). 

I'm a bit concerned about the non-standard way the Lit-mixin (i.e. the open WC stuff under the hood) currently handles the registration. A polyfill is needed even for browsers that actually DO support the SCER features as specced - like Chrome and Safari. 

If I use this in my design system library to get crossbrowser support today, will I always need the polyfill for it to function? In other words what's the path for this mixin regarding the removal of the polyfill?

Thanks!

**Reply by piyush8098** [reply]
I wonder if anyone has any update on the proposal (https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md) based on which the polyfill was written. This proposal is 6 years old; we still don't see any browser implementing it.

Moreover there are a couple of open questions in it. I wonder if those has been addressed yet?
Overall, does anyone has any update about this old proposal at all? Will it make it to the browser of we need to forever keep using the polyfill (which is not ideal for production grade applications)

**Reply by tomherni** [reply]
> I wonder if anyone has any update on the proposal (https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md) based on which the polyfill was written. This proposal is 6 years old; we still don't see any browser implementing it.
> 
> Moreover there are a couple of open questions in it. I wonder if those has been addressed yet? Overall, does anyone has any update about this old proposal at all? Will it make it to the browser of we need to forever keep using the polyfill (which is not ideal for production grade applications)

The spec advanced and there is an open PR to update the polyfill. It's blocked on review at the moment.

Browsers already shipped scoped registries, except for Firefox where it's still in the Nightly. [MDN reference](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/CustomElementRegistry).

**Reply by LorenzKahl** [reply]
There seems to be some movements here: https://wpt.fyi/results/custom-elements/registries?label=master&label=experimental&aligned&view=interop&q=label%3Ainterop-2026-scoped-custom-element-registries

**Reply by piyush8098** [reply]
> > I wonder if anyone has any update on the proposal (https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md) based on which the polyfill was written. This proposal is 6 years old; we still don't see any browser implementing it.
> > Moreover there are a couple of open questions in it. I wonder if those has been addressed yet? Overall, does anyone has any update about this old proposal at all? Will it make it to the browser of we need to forever keep using the polyfill (which is not ideal for production grade applications)
> 
> The spec advanced and there is an open PR to update the polyfill. It's blocked on review at the moment.
> 
> Browsers already shipped scoped registries, except for Firefox where it's still in the Nightly. [MDN reference](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/CustomElementRegistry).

Thanks Tom! It's good news that Polyfill has a PR pending for review.
In regard with browsers shipping scoped CustomElementRegistry, I have noticed a bug with the way browsers implement it. I have raised a bug (https://issues.chromium.org/issues/558819239) with Chromium which seems to have been accepted and elevated as P1. The issue exists with Safari and edge too.

**Reply by piyush8098** [reply]
> There seems to be some movements here: https://wpt.fyi/results/custom-elements/registries?label=master&label=experimental&aligned&view=interop&q=label%3Ainterop-2026-scoped-custom-element-registries

Oh thanks for sharing it!! These browsers do have a bug with this feature. Which I have [reported](https://issues.chromium.org/issues/558819239)

**Reply by sorvell** [reply]
It’s not a bug. That’s per spec. The element was created with the global registry.

**Reply by piyush8098** [reply]
> It’s not a bug. That’s per spec. The element was created with the global registry.

I believe this is a bug. Previously, when a custom element was added using shadowRoot.appendChild(document.createElement('cust-element')) followed by registry.upgrade(shadowRoot), it worked as expected up to Chrome 149. However, the behaviour changed in later Chrome versions and the upgrade no longer occurs as expected.

Google has been able to reproduce the issue on their end as well. As shown in the attached screenshot, they successfully replicated the steps I outlined in my bug report comments. Given the reproducibility and impact of the issue, it has now been escalated and logged as a P1 ticket.

<img width="1470" height="956" alt="558819239-Good" src="https://github.com/user-attachments/assets/8f5d7be9-45c2-44c4-aed6-eaabae0efd3f" />

<img width="1470" height="956" alt="558819239-Bad" src="https://github.com/user-attachments/assets/dbd6c759-6137-4866-ba7b-1edc6fa3c528" />

**Reply by sorvell** [reply]
Yeah, it's a little confusing but what you showed is not a bug. 

TLDR: You'll get the behavior you expect via `const appendedEl = document.createElement('my-el', {customElementRegistry: scopedRegistry});` or `document.createElement('my-el', {customElementRegistry: null});` 

Let me explain:

The spec changed so you can't use older Chrome as a baseline. See https://github.com/whatwg/html/issues/10854 for more info. The Chrome QA reviewer doesn't have the context to know this. This will be corrected when the implementer gets to it. Fwiw, I suggest posting a reproduction along with steps to repro on Chrome bugs.

Here's what's happening:

When an element is created, it's `customElementRegistry` is set. This can *only* be changed (by calling `registry.initialize` on it or a subtree it's in) if it's created with a `null` registry. For back compat `const appendedEl = document.createElement('my-el')` creates the element in the document registry so `appendedEl.customElement === window.customElements`. Appending that element to a shadowRoot with a different registry doesn't change the registry since it didn't start as `null`. So you either need to create the element with the registry you want or a null registry. If it's null, then you can initialize it with `registry.initialize` or attach it to a subtree with a registry. The `createElement` function's 2nd "options" argument is where you specify the registry: `{customElementRegistry: scopedRegistry}`.

**Reply by piyush8098** [reply]
Thank you @sorvell for taking the time to write down the explanation! Appreciate it.
